### PR TITLE
fixed all commandline tests for mono

### PR DESCRIPTION
**Pull request by zhili1208**
NuGet.CommandLine test fix for mono

### CODE REVIEW

**Comment by joelverhagen** [comment]
Please rebase first.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs [hunk 1]
```diff
@@ -559,11 +560,16 @@ public void InitCommand_Fail_SourceIsInvalid()
 
                 // Assert
                 var expectedMessage = string.Format(NuGetResources.Path_Invalid, invalidPath);
+
+                if (RuntimeEnvironmentHelper.IsMono && !RuntimeEnvironmentHelper.IsWindows)
+                {
+                    expectedMessage = string.Format(NuGetResources.InitCommand_FeedIsNotFound, invalidPath);
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs:567]: Leave comment that `foo|<>|bar` is a valid directory name in Unix.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs [hunk 2]
```diff
@@ -559,11 +560,16 @@ public void InitCommand_Fail_SourceIsInvalid()
...
+                }
                 Util.VerifyResultFailure(result, expectedMessage);
             }
         }
 
-        [Fact]
+        [WindowsNTFact]
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs:573]: Create another test that verifies that `foo|<>|bar` is created on Unix.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs [hunk 3]
```diff
@@ -523,7 +522,7 @@ public void InstallCommand_OptOutMessage(string configFileName)
                 var r = CommandRunner.Run(
                     nugetexe,
                     proj1Directory,
-                    "install " + configFileName + " -Source " + repositoryPath + @" -ConfigFile ..\my.config -RequireConsent -Verbosity detailed",
+                    "install " + configFileName + " -Source " + repositoryPath + $@" -ConfigFile ..{Path.AltDirectorySeparatorChar}my.config -RequireConsent -Verbosity detailed",
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: Should be `DirectorySeparatorChar`? Alt seems wrong. dbl check plz

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs [hunk 4]
```diff
@@ -588,7 +587,7 @@ public void InstallCommand_NoOptOutMessage(string configFileName)
                 var r = CommandRunner.Run(
                     nugetexe,
                     proj1Directory,
-                    "install " + configFileName + " -Source " + repositoryPath + @" -ConfigFile ..\my.config",
+                    "install " + configFileName + " -Source " + repositoryPath + $@" -ConfigFile ..{Path.AltDirectorySeparatorChar}my.config",
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: same

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetListCommandTest.cs [hunk 5]
```diff
@@ -937,7 +962,7 @@ public void ListCommand_InvalidInput_V3_NonExistent(string invalidInput)
                 "The run did not fail as desired. Simply got this output:" + result.Item2);
 
             Assert.True(
-                result.Item3.Contains("An error occurred while sending the request."),
+                result.Item3.Contains("Unable to load the service index for source https://invalid-2a0358f1-88f2-48c0-b68a-bb150cac00bd.org/v3/index.json."),
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetListCommandTest.cs]: Use `invalidInput` local variable in your assert.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs [hunk 6]
```diff
@@ -1673,13 +1677,17 @@ public void PackCommand_ReferencedProjectWithJsonFileWithMsbuild14()
     }
   }
 }");
-
+                var version = "14";
+                if (RuntimeEnvironmentHelper.IsMono && !RuntimeEnvironmentHelper.IsWindows)
+                {
+                    version = "15.0";
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs:1791]: Add issue to track testing `15.0` on Windows and link to the issue in each of these `IsMono` conditions.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs [hunk 7]
```diff
@@ -4244,7 +4255,14 @@ public void PackCommand_MissingPackageCausesError()
                 Assert.Equal(1, r.Item1);
 
                 // Assert
-                Assert.Contains("Unable to find 'doesNotExist.1.1.0.nupkg'.", r.Item3);
+                if (RuntimeEnvironmentHelper.IsMono && !RuntimeEnvironmentHelper.IsWindows)
+                {
+                    Assert.Contains("Failed to build 'proj1.csproj'.", r.Item3);
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs:4597]: Do we still need this? We spoke offline about this one.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs [hunk 8]
```diff
@@ -79,7 +80,7 @@ public void PushCommand_PushToV3FileSystemSource()
                 // Assert
                 Assert.Equal(0, result.Item1);
                 var basename = string.Format("{0}.{1}.", packageId, version);
-                var baseFolder = string.Format("{0}\\{1}\\", packageId, version);
+                var baseFolder = string.Format("{0}" + Path.DirectorySeparatorChar + "{1}" + Path.DirectorySeparatorChar, packageId, version);
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs]: nit: use `Path.Combine`.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs [hunk 9]
```diff
@@ -603,7 +604,7 @@ public void PushCommand_PushToServerWithInvalidRedirectionLocation()
         // Regression test for the bug that "nuget.exe push" will retry forever instead of asking for
         // user's password when NuGet.Server uses Windows Authentication.
         [Fact(Skip = "TODO: reconstruct faked response headers which won't crash HttpClient. " +
-            "Using real server, same scenario works fine")]
+            "Using real server, same sceanrio works fine")]
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs]: misspell

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs [hunk 10]
```diff
@@ -1763,7 +1763,7 @@ public void PushCommand_APIV2Package_Endpoint()
                         "push",
                         packageFileName,
                         "-Source",
-                        server.Uri + "/api/v2/Package"
+                        server.Uri + "api/v2/Package"
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:1766]: What part of Mono is failing on the double slash? This test seems fine, but we should understand the root cause here. The different in `//` handling seems a bit scary.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs [hunk 11]
```diff
@@ -481,11 +481,10 @@ public void RestoreCommand_FromSolutionFileWithMsbuildPath()
             var nugetexe = Util.GetNuGetExePath();
 
 
-            var msbuildPath = Util.GetMsbuildPathOnWindows();
-            var os = Environment.GetEnvironmentVariable("OSTYPE");
-            if (RuntimeEnvironmentHelper.IsMono && os != null && os.StartsWith("darwin"))
+            var msbuildPath = @"C:\Program Files (x86)\MSBuild\14.0\Bin";
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs]: We should still call `Util.GetMsbuildPathOnWindows()` right?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs [hunk 12]
```diff
@@ -540,11 +539,10 @@ public void RestoreCommand_FromSolutionFileWithMsbuildPathAndMsbuildVersion()
...
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs]: same

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetSourcesCommandTest.cs [hunk 13]
```diff
@@ -1,7 +1,4 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetSourcesCommandTest.cs:1]: keep license

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetSourcesCommandTest.cs [hunk 14]
```diff
@@ -366,12 +367,13 @@ public void TestVerbosityQuiet_DoesNotShowInfoMessages()
                     "-Verbosity",
                     "Quiet"
                 };
+                string root = RuntimeEnvironmentHelper.IsMono && !RuntimeEnvironmentHelper.IsWindows ? Environment.GetEnvironmentVariable("HOME") : @"c:\";
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetSourcesCommandTest.cs]: Maybe create a util here `Directory.GetDirectoryRoot(Directory.GetCurrentDirectory())`? Or maybe use a temp dir? Or can we use home dir on both platforms? Ideally Windows vs. Mono test code is the same.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/PortReserver.cs [hunk 15]
```diff
@@ -22,6 +23,7 @@ namespace NuGet.CommandLine.Test
     public class PortReserver : IDisposable
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/PortReserver.cs:23]: Can we just drop-in replace this with:
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Test.Server/PortReserver.cs

Comment (joelverhagen) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/PortReserver.cs:23]: Actually, there are significant API changes here. Can you file an issue to consolidate this code?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs [hunk 16]
```diff
@@ -1,19 +1,20 @@
-﻿using System;
+﻿using Moq;
+using Newtonsoft.Json.Linq;
+using NuGet.Frameworks;
+using NuGet.Packaging;
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs]: NuGet should go after System. Same with Moq.

**Review by joelverhagen — APPROVED** [review]
Address comments then :shipit:

**Comment by zhili1208** [comment]
https://github.com/NuGet/NuGet.Client/commit/16ae864cb5a4d6e051a369d6eae682643d4088a6

**Pull request closed without merging**